src/mam4xx/coagulation.hpp

```diff
@@ -59,7 +64,1120 @@ class Coagulation {
+// - coag pair 1: aitken + accumulation -> accumulation
+// - coag pair 2: pca    + accumulation -> accumulation
+// - coag pair 3: aitken + pca          -> pca
+//
+// Sorted by mass source:
+//  - From aitken mode:
+//      - coag pair 1: aitken + accumulation -> accumulation
+//      - coag pair 3: aitken + pca          -> pca
+//  - From pca mode:
+//      - coag pair 2: pca    + accumulation -> accumulation
+//--------------------------------------------------------------------------------
+// Numerical treatment:
+//
+// The transfer amounts are calculated using as an exponential decay of
+// the initial mass mixing ratios,
+// where the decay constant is calculated using the average (over one timestep)
+// number mixing ratios for each mode
+//
+// The mass transfer calculations are first-order accurate in time,
+// because the mass transferred out of a mode does not
+// include any mass transferred in during the time step.
+// With this approach, the ordering is not important, but the mass transfer
+// calculations are done in the reverse order of the number loss calculations
+// --------------------------------------------------------
+KOKKOS_INLINE_FUNCTION
+void mam_coag_aer_update(
+    Real ybetaij3[Coagulation::max_coagpair], Real deltat,
+    Real qnum_tavg[AeroConfig::num_modes()],
+    Real qaer_bgn[AeroConfig::num_aerosol_ids()][AeroConfig::num_modes()],
+    Real qaer_end[AeroConfig::num_aerosol_ids()][AeroConfig::num_modes()],
+    Real qaer_del_coag_out[AeroConfig::num_aerosol_ids()]
+                          [Coagulation::max_agepair]) {
+
+  const int num_aer = AeroConfig::num_aerosol_ids();
+  const int num_mode = AeroConfig::num_modes();
+  // --------------------------------------------------------------------
+  // Initialize the array that will be passed onto aging
+  // --------------------------------------------------------------------
+  for (int ispec = 0; ispec < num_aer; ++ispec) {
+    for (int imode = 0; imode < num_mode; ++imode) {
+      qaer_del_coag_out[ispec][imode] = 0.0;
+    }
+  }
+
+  // --------------------------------------------------------------------
+  //  Mass transfer out of aitken mode. Two coag pairs are involved:
+  // - coag pair 1: aitken + accumulation -> accumulation
+  // - coag pair 3: aitken + pca          -> pca
+  // --------------------------------------------------------------------
+  // Calculate the rate of mass transfer into different destination modes and
+  // the sum over all modes
+
+  const int nacc = static_cast<int>(ModeIndex::Accumulation);
+  const int npca = static_cast<int>(ModeIndex::PrimaryCarbon);
+  const int nait = static_cast<int>(ModeIndex::Aitken);
+
+  const Real bijqnumj1 = haero::max(0.0, ybetaij3[0] * qnum_tavg[nacc]);
+  const Real bijqnumj2 = haero::max(0.0, ybetaij3[2] * qnum_tavg[npca]);
+  Real decay_const = bijqnumj1 + bijqnumj2;
+
+  constexpr float epsilonx2 = std::numeric_limits<float>::epsilon() * 2.0;
```

Comment:
Are we allow to use std functions in CUDA?

Reply:
Well, I don't know that this is supposed to work--the `std` namespace is not available to CUDA C++ as far as I know. Can someone try to build this on a GPU workstation available to them?

Reply:
The std::numeric_limits() compiles and the tests pass with Cuda 11.2 and GCC 10.3 on Linux.

Reply:
Well, that's interesting, though it doesn't give me a very warm and fuzzy feeling! I guess we'll find out how intentional this is as we build on more exotic platforms.

Thanks for testing, James.

Reply:
I believe there is an equivalent version of std::numeric_limits<float>::epsilon() in kokkos.